src/rpc/blockchain.cpp

```diff
@@ -1916,6 +1920,289 @@ static UniValue savemempool(const JSONRPCRequest& request)
     return NullUniValue;
 }
 
+//! Search for a given set of pubkey scripts
+bool FindScriptPubKey(std::atomic<int>& scan_progress, const std::atomic<bool>& should_abort, int64_t& count, CCoinsViewCursor* cursor, const std::set<CScript>& needles, std::map<COutPoint, Coin>& out_results) {
+    scan_progress = 0;
+    count = 0;
+    while (cursor->Valid()) {
+        COutPoint key;
+        Coin coin;
+        if (!cursor->GetKey(key) || !cursor->GetValue(coin)) return false;
+        if (++count % 8192 == 0) {
+            boost::this_thread::interruption_point();
+            if (should_abort) {
+                // allow to abort the scan via the abort reference
+                return false;
+            }
+        }
+        if (count % 256 == 0) {
+            // update progress reference every 256 item
+            uint32_t high = 0x100 * *key.hash.begin() + *(key.hash.begin() + 1);
+            scan_progress = (int)(high * 100.0 / 65536.0 + 0.5);
```

Comment:
Nit: code for this seems like it's cropping up in a few different places, might be nice to have an abstraction for it.
```
 $ git grep -C 1 "/ 65536" | cat

src/index/txindex.cpp-                (static_cast<uint32_t>(*(txid.begin() + 1)) << 0);
src/index/txindex.cpp:            int percentage_done = (int)(high_nibble * 100.0 / 65536.0 + 0.5);
src/index/txindex.cpp-
--
src/txdb.cpp-                uint32_t high = 0x100 * *key.second.begin() + *(key.second.begin() + 1);
src/txdb.cpp:                int percentageDone = (int)(high * 100.0 / 65536.0 + 0.5);
src/txdb.cpp-                uiInterface.ShowProgress(_("Upgrading UTXO database"), percentageDone, true);
```

Reply:
IMO this should be done outside of this PR